### PR TITLE
Format low level partition assignment logging

### PR DESCRIPTION
**Pull request by shrinandthakkar**
Formatted logs which shortens the list of topic-partition mappings by merging partitions of the same topic together. 
e.g,
topic1-0, topic1-1, topic2-0 -> topic1:[0-1], topic2:[0]

### CODE REVIEW

**Review by vmaheshw — CHANGES_REQUESTED** [review]
Thanks Shrinand for the change. Please add unit-tests. Also, add the after output logs to show how you verified this work.

File: datastream-utils/src/main/java/com/linkedin/datastream/common/LogUtils.java [hunk 1]
```diff
@@ -62,4 +66,30 @@ public static String logNumberArrayInRange(List<Integer> list) {
...
     }
   }
+
+  /**
+   * Shortening the list of topic-partition mappings by merging partitions of the same topic together. e.g.
+   * topic1-0, topic1-1, topic2-0 -> topic1:[0-1], topic2:[0]
+   * @param list list of strings to generate logging String for
+   * @return compacted String that merges partitions per topic
+   */
+  public static String logOrganizeTopicPartitionMappings(List<String> list) {
```
Comment (vmaheshw) [inline datastream-utils/src/main/java/com/linkedin/datastream/common/LogUtils.java:76]: use meaningful name instead of `list`

File: datastream-utils/src/main/java/com/linkedin/datastream/common/LogUtils.java [hunk 2]
```diff
@@ -62,4 +66,30 @@ public static String logNumberArrayInRange(List<Integer> list) {
...
+    Map<String, List<Integer>> topicPartitionsV2 = new HashMap<>();
+    try {
+      for (String partition : list) {
+        String[] tokens = partition.split("-");
```
Comment (vmaheshw) [inline datastream-utils/src/main/java/com/linkedin/datastream/common/LogUtils.java:80]: can you reuse [code](https://github.com/linkedin/brooklin/blob/d12144a4dad2bb973ca7fcc9c6088c5c46a49172/datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/TopicPartitionUtil.java) to split topic/partition name?

File: datastream-utils/src/main/java/com/linkedin/datastream/common/LogUtils.java [hunk 3]
```diff
@@ -62,4 +66,30 @@ public static String logNumberArrayInRange(List<Integer> list) {
       return list.toString();
     }
   }
+
+  /**
+   * Shortening the list of topic-partition mappings by merging partitions of the same topic together. e.g.
+   * topic1-0, topic1-1, topic2-0 -> topic1:[0-1], topic2:[0]
+   * @param list list of strings to generate logging String for
+   * @return compacted String that merges partitions per topic
+   */
```
Comment (vmaheshw) [inline datastream-utils/src/main/java/com/linkedin/datastream/common/LogUtils.java:75]: Suggestion: Name of the method can change to `logSummarizedTopicPartitionsMapping`

File: datastream-utils/src/main/java/com/linkedin/datastream/common/LogUtils.java [hunk 4]
```diff
@@ -62,4 +66,30 @@ public static String logNumberArrayInRange(List<Integer> list) {
       return list.toString();
     }
   }
+
+  /**
+   * Shortening the list of topic-partition mappings by merging partitions of the same topic together. e.g.
+   * topic1-0, topic1-1, topic2-0 -> topic1:[0-1], topic2:[0]
+   * @param list list of strings to generate logging String for
```
Comment (vmaheshw) [inline datastream-utils/src/main/java/com/linkedin/datastream/common/LogUtils.java:73]: nit: String to string

File: datastream-utils/src/main/java/com/linkedin/datastream/common/LogUtils.java [hunk 5]
```diff
@@ -62,4 +66,30 @@ public static String logNumberArrayInRange(List<Integer> list) {
...
+   * @return compacted String that merges partitions per topic
+   */
+  public static String logOrganizeTopicPartitionMappings(List<String> list) {
+    Map<String, List<Integer>> topicPartitionsV2 = new HashMap<>();
```
Comment (vmaheshw) [inline datastream-utils/src/main/java/com/linkedin/datastream/common/LogUtils.java:77]: the variable name does not have to be topicPartitionsV2, since it is a utility. `topicPartitionsMap` will work.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java [hunk 6]
```diff
@@ -297,8 +297,8 @@ public StickyPartitionAssignmentStrategy(Optional<Integer> maxTasks, Optional<In
   public Map<String, Set<DatastreamTask>> movePartitions(Map<String, Set<DatastreamTask>> currentAssignment,
       Map<String, Set<String>> targetAssignment, DatastreamGroupPartitionsMetadata partitionsMetadata) {
 
-    LOG.info("Move partition, current assignment: {}, target assignment: {}, all partitions: {}", currentAssignment,
-        targetAssignment, partitionsMetadata.getPartitions());
+    LOG.info("Move partition, current assignment: {}, target assignment: {}, partitions meta: {}", currentAssignment,
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java:300]: nit: meta to metadata

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java [hunk 7]
```diff
@@ -270,8 +270,8 @@ public StickyPartitionAssignmentStrategy(Optional<Integer> maxTasks, Optional<In
       }).collect(Collectors.toSet());
       newAssignment.put(instance, newAssignedTask);
     });
-    LOG.info("new assignment info, assignment: {}, all partitions: {}", newAssignment,
-        datastreamPartitions.getPartitions());
+    LOG.info("new assignment info, assignment: {}, partitions meta: {}", newAssignment,
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java:273]: nit: meta to metadata

File: datastream-server/src/main/java/com/linkedin/datastream/server/DatastreamTaskImpl.java [hunk 8]
```diff
@@ -479,6 +479,8 @@ public String toString() {
       } catch (NumberFormatException e) {
         LOG.error(e.getMessage());
       }
+    } else { // this will be called for low level kafka connector.
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/DatastreamTaskImpl.java:482]: This should not be in else block. Check line 472.

`// When DatastreamTaskImpl is created using constructor that passes _partitions, _partitionsV2 is also populated.`

File: datastream-utils/src/main/java/com/linkedin/datastream/common/LogUtils.java (hunk 2 above)
Comment (shrinandthakkar) [reply datastream-utils/src/main/java/com/linkedin/datastream/common/LogUtils.java]: Couldn't reuse that function from its original package, as there was a modular circular dependency. So, moved it to common module.

File: datastream-server/src/main/java/com/linkedin/datastream/server/DatastreamTaskImpl.java (hunk 8 above)
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/DatastreamTaskImpl.java:482]: my bad !! makes sense.

File: datastream-utils/src/test/java/com/linkedin/datastream/common/TestLogUtils.java [hunk 9]
```diff
@@ -31,4 +31,23 @@ public void testLogNumberArrayInRange() throws Exception {
...
+    // don't modify the original input list
+    Assert.assertEquals(input.get(0), "t-3");
+
+    input = Arrays.asList("t-1", "t-5", "t-6", "t-4", "t-2", "t-3", "t1-1", "t1-5", "t1-6", "t1-4", "t1-2", "t1-3");
+    Assert.assertEquals(LogUtils.logSummarizedTopicPartitionsMapping(input), "t:[1-6], t1:[1-6]");
+
+    input = Arrays.asList("t-1", "t-5", "t-6", "t-4", "t-2", "t-");
+    Assert.assertEquals(LogUtils.logSummarizedTopicPartitionsMapping(input), "t-1,t-5,t-6,t-4,t-2,t-");
+
+    Assert.assertEquals(LogUtils.logSummarizedTopicPartitionsMapping(null), "[]");
```
Comment (vmaheshw) [inline datastream-utils/src/test/java/com/linkedin/datastream/common/TestLogUtils.java:51]: add test case for empty list.

**Review by vmaheshw — APPROVED** [review]
LGTM. Please add additional tests.

File: datastream-utils/src/test/java/com/linkedin/datastream/common/TestLogUtils.java [hunk 10]
```diff
@@ -31,4 +31,23 @@ public void testLogNumberArrayInRange() throws Exception {
 
     Assert.assertEquals(LogUtils.logNumberArrayInRange(null), "[]");
   }
+
+  @Test
+  public void testLogSummarizedTopicPartitionsMapping() throws Exception {
+    List<String> input = Arrays.asList("t-1", "t-3", "t-6", "t-4", "t-7", "t-12", "t-9", "t-10", "t-8");
```
[the remark this answers is not in the document]
Comment (shrinandthakkar) [reply datastream-utils/src/test/java/com/linkedin/datastream/common/TestLogUtils.java:37]: could there be an actual case like that? 
> Eg, "t-0, t-1, t--2" --> "t:[0-2]"
Since reusing this [code](https://github.com/linkedin/brooklin/blob/master/datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/TopicPartitionUtil.java) would give output like,
> "t:[0-1], t-:[2]"

[the remark this answers is not in the document]
Comment (vmaheshw) [reply datastream-utils/src/test/java/com/linkedin/datastream/common/TestLogUtils.java:37]: the topic name can be topic-abc-def-ghi-xyz

File: datastream-utils/src/test/java/com/linkedin/datastream/common/TestLogUtils.java [hunk 11]
```diff
@@ -31,4 +31,27 @@ public void testLogNumberArrayInRange() throws Exception {
...
+    Assert.assertEquals(LogUtils.logSummarizedTopicPartitionsMapping(input), "t:[1, 3-4, 6-10, 12]");
+
+    input = Arrays.asList("t-3", "t-1", "t-4", "t-6", "t-7", "t-8", "t-9", "t-9", "t-9", "t-10", "t-11", "t-12", "t1-9", "t1-10", "t1-11", "t1-12");
+    Assert.assertEquals(LogUtils.logSummarizedTopicPartitionsMapping(input), "t:[1, 3-4, 6-12], t1:[9-12]");
+    // don't modify the original input list
+    Assert.assertEquals(input.get(0), "t-3");
+
+    input = Arrays.asList("t-1", "t-5", "t-6", "t-4", "t-2", "t-3", "t1-1", "t1-5", "t1-6", "t1-4", "t1-2", "t1-3");
+    Assert.assertEquals(LogUtils.logSummarizedTopicPartitionsMapping(input), "t:[1-6], t1:[1-6]");
```
Comment (vmaheshw) [inline datastream-utils/src/test/java/com/linkedin/datastream/common/TestLogUtils.java:46]: one thing for the test: logSummarizedTopicPartitionsMapping you stream the map on keyset(), I don't think keyset() guarantee ordering, so the topics can get reordered and the string may be correct, but the test can fail

May be: what you can do is, split the output string in test by ", " and then sort the list and compare two list

https://stackoverflow.com/questions/1882762/is-the-java-hashmap-keyset-iteration-order-consistent

Comment (shrinandthakkar) [reply datastream-utils/src/test/java/com/linkedin/datastream/common/TestLogUtils.java:46]: good point

**Review by jzakaryan — APPROVED** [review]
looks good to me

File: datastream-utils/src/test/java/com/linkedin/datastream/common/TestLogUtils.java [hunk 12]
```diff
@@ -31,4 +34,34 @@ public void testLogNumberArrayInRange() throws Exception {
 
     Assert.assertEquals(LogUtils.logNumberArrayInRange(null), "[]");
   }
+
+  @Test
+  public void testLogSummarizedTopicPartitionsMapping() throws Exception {
+    List<String> input = Arrays.asList("t-1", "t-3", "t-6", "t-4", "t-7", "t-12", "t-9", "t-10", "t-8");
+    Assert.assertEquals(LogUtils.logSummarizedTopicPartitionsMapping(input), "t:[1, 3-4, 6-10, 12]");
+
+    input = Arrays.asList("t-3", "t-1", "t-4", "t-6", "t-7", "t-8", "t-9", "t-9", "t-9", "t-10", "t-11", "t-12", "t1-9", "t1-10", "t1-11", "t1-12");
```
Comment (vmaheshw) [inline datastream-utils/src/test/java/com/linkedin/datastream/common/TestLogUtils.java:43]: nit picking: can you extract method out of this for easy readability?

**Review by vmaheshw — APPROVED** [review]

**Review by vmaheshw — APPROVED** [review]

File: datastream-utils/src/test/java/com/linkedin/datastream/common/TestLogUtils.java [hunk 13]
```diff
@@ -31,4 +35,43 @@ public void testLogNumberArrayInRange() throws Exception {
...
+
+    Assert.assertEquals(LogUtils.logSummarizedTopicPartitionsMapping(null), "[]");
+    Assert.assertEquals(LogUtils.logSummarizedTopicPartitionsMapping(Collections.emptyList()), "[]");
+  }
+
+  /**
+   * Get all topic partition mappings in a set tokenized by ", "
+   * @param topicPartitionMapping topic partition string to be tokenized
+   */
+  public Set<String> getTokenizedTopicPartitionMappings(String topicPartitionMapping) {
```
Comment (vmaheshw) [inline datastream-utils/src/test/java/com/linkedin/datastream/common/TestLogUtils.java:73]: remove the public part and javadoc.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java [hunk 14]
```diff
@@ -270,8 +270,8 @@ public StickyPartitionAssignmentStrategy(Optional<Integer> maxTasks, Optional<In
       }).collect(Collectors.toSet());
       newAssignment.put(instance, newAssignedTask);
     });
-    LOG.info("new assignment info, assignment: {}, all partitions: {}", newAssignment,
-        datastreamPartitions.getPartitions());
+    LOG.info("new assignment info, assignment: {}, partitions metadata: {}", newAssignment,
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java:273]: split it in two lines for better readability, after checking in the logs.

**Review by jzakaryan — APPROVED** [review]

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java [hunk 15]
```diff
@@ -297,8 +297,8 @@ public StickyPartitionAssignmentStrategy(Optional<Integer> maxTasks, Optional<In
   public Map<String, Set<DatastreamTask>> movePartitions(Map<String, Set<DatastreamTask>> currentAssignment,
       Map<String, Set<String>> targetAssignment, DatastreamGroupPartitionsMetadata partitionsMetadata) {
 
-    LOG.info("Move partition, current assignment: {}, target assignment: {}, all partitions: {}", currentAssignment,
-        targetAssignment, partitionsMetadata.getPartitions());
+    LOG.info("Move partition, current assignment: {},\ntarget assignment: {},\nall datastream partitions: {}", currentAssignment,
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java:300]: Please split it into 3 lines, for grep/filtering based on the class name.

**Review by vmaheshw — APPROVED** [review]

**Review by jzakaryan — APPROVED** [review]
